### PR TITLE
TypeScript + HMR

### PR DESCRIPTION
**Pull request by janklimo**
#### Changes

1. Added TypeScript support with a tiny demo component
1. Added a CI build step to perform TS type-checking. `yarn types-watch` can be run locally to compile TS code in watch mode. However, setting up your editor to perform it is preferable.
1. Updated readme.
1. Added polyfill support with `@babel/preset-env`.
1. Restricted chunk-splitting to production only to make HMR possible.
1. Added support for HMR.

### CODE REVIEW

**Comment by hiattp** [comment]
Cool! Will dig into this end of next week after I wrap up my milestone tickets

File: app/javascript/components/Counter.tsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+  return (
+    <>
+      <p>You clicked {count} times!</p>
+      <button onClick={() => setCount(count + 1)}>Click me</button>
+    </>
+  );
+}
+
+export default hot(Counter);
```
Comment (hiattp) [inline app/javascript/components/Counter.tsx]: I'd rather not have example code like this, we should be able to put a snippet in the readme or do a link to a gist or something so we don't have to remember to remove this on every new application

Comment (janklimo) [reply app/javascript/components/Counter.tsx]: @hiattp just wanted an example showing a HMR-ready export, though on a second thought, it's enough to mention in readme 👍

Comment (janklimo) [reply app/javascript/components/Counter.tsx]: @hiattp updated

File: README.md [hunk 2]
```diff
@@ -168,40 +164,67 @@ Authentication concerns (your typical `Devise` configuration) are handled by the
...
 
 ```
 heroku buildpacks:clear
 heroku buildpacks:set heroku/nodejs
 heroku buildpacks:add heroku/ruby
 ```
 
+### TypeScript
+
+TypeScript is supported out of the box. See the example in `app/javascript`.
```
Comment (hiattp) [inline README.md]: Did we remove the example this refers to?

File: app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -20,5 +19,6 @@
   <body>
     <%= render partial: 'layouts/flash' %>
     <%= yield %>
+    <%= javascript_pack_tag 'application' %>
```
Comment (hiattp) [inline app/views/layouts/application.html.erb]: We don’t want this in the head?

File: README.md (hunk 2 above)
Comment (janklimo) [reply README.md]: well that example is TS now, would you prefer to change it back?

File: app/views/layouts/application.html.erb (hunk 3 above)
Comment (janklimo) [reply app/views/layouts/application.html.erb]: Good point, it should stay in the head (this was needed by the example I removed).

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -10,7 +10,7 @@
     <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
     <%= render partial: 'layouts/google_analytics' %>
     <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
-    <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_packs_with_chunks_tag 'application', 'data-turbolinks-track': 'reload' %>
```
Comment (janklimo) [inline app/views/layouts/application.html.erb]: @hiattp moved this back to the head and fixed the tag to use chunks (it was missing from a previous PR)

**Pull request merged**